leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/Lwm2mClientEndpointContextMatcher.java

```diff
@@ -0,0 +1,71 @@
+/*******************************************************************************
+ * Copyright (c) 2017 Bosch Software Innovations GmbH and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial implementation
+ ******************************************************************************/
+package org.eclipse.leshan.core.californium;
+
+import org.eclipse.californium.elements.auth.X509CertPath;
+import org.eclipse.leshan.core.request.Identity;
+
+import javax.security.auth.x500.X500Principal;
+import java.security.Principal;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.List;
+
+/**
+ * LWM2M principal based endpoint context matcher for LWM2M Clients.
+ * 
+ * Matches DTLS based on the used principal. Requires unique credentials.
+ * 
+ * There is separation between LWM2M Client and Server side EndpointContextMatchers as when client is performing
+ * initial connection it doesn't know real peer identity due to flexibility of certificate usage configuration.
+ *
+ * With certificate usage the server identity may be the intermediate CA or root CA. For this reason the EndpointContext
+ * matcher is relaxed to allow matching to other parts of the chain. This can be done as before the context matcher
+ * there is DTLS Handshake which has verified that certificate usage and chain is expected.
+ *
+ * When server is receiving DTLS connection it always knows the whole peer identity as DTLS handshake is performed first
+ * before any message is being sent to client and then EndpointContext is built with real client information. In server
+ * more lightweight and strict matching can be done there and there is no need to inspect the chain.
+ * 
+ * For more detailed about why this is needed, see
+ * <a href="https://github.com/eclipse/leshan/wiki/LWM2M-Observe#for-dtls">LWM2M-Observe wiki page</a>
+ */
+public class Lwm2mClientEndpointContextMatcher extends Lwm2mEndpointContextMatcher {
```

Comment:
I don't know why but  LWM2M in class name is written like this : `LwM2m` :exploding_head: 

I can understand this way to write LWM2M does not make so much sense, so we could eventually change it if we think this is worth (but for the whole project in another PR).

So there is typo in `Lwm2mEndpointContextMatcher` and `Lwm2mEndpointContextMatcher`.

But maybe we don't even need a dedicated class a simple inline class is maybe enough ?

If we decide to keep the class : 
- I think the javadoc should updated too. (The javadoc explanation should be something like this : https://github.com/eclipse/leshan/pull/903#issuecomment-706158960
- I think that it should directly implement `PrincipalEndpointContextMatcher`
- `Lwm2mEndpointContextMatcher` should probably be renamed in `LwM2mServerEndpointContextMatcher`

Reply:
Removed the extra class and made it anonymous class.